lib/os/cbprintf_packaged.c

```diff
@@ -32,13 +32,13 @@ static inline bool ptr_in_rodata(const char *addr)
 #elif defined(CONFIG_ARC) || defined(CONFIG_ARM) || defined(CONFIG_X86) \
 	|| defined(CONFIG_RISCV) || defined(CONFIG_ARM64) \
 	|| defined(CONFIG_NIOS2)
-	extern char _image_rodata_start[];
-	extern char _image_rodata_end[];
-#define RO_START _image_rodata_start
-#define RO_END _image_rodata_end
+	extern char __rom_start_start[];
```

Comment:
> I'm confused about this patch. Let's take this line as an example.
> 
> This line changes RO_START from _image_rodata_start to _rom_start_start on most architectures.

I think you're right at multiple things.
First, the part of aligning linker symbol is still at first stage, as I have focused on the CMake part to provide enough of the CMake linker feature to see that this direction is indeed feasible, and as part of that I have only done the necessary cleanup to get `hello world` running when using both existing infrastructure, but also with generated linker scripts (ld and scatter).

In scatter files we cannot control symbol ourselves, instead armlink will provide linker defined symbols, based on section names. This is why a steering file is used to link the names we uses to names provided by armlink.
In general I prefer that we (Zephyr) defines the names we want to use, and not the linker we use.

Most places, Zephyr also define linker symbols around output sections which almost follows the name of the section.
The `__image_rom_X` and `__image_rodata_X` being exceptions.
Now, handling of such groups are still wip, and can indirectly be seen by this commit message: https://github.com/zephyrproject-rtos/zephyr/pull/36140/commits/0dd656f18b65072d0562786e4a3b40aaea76d478 but I should probably have made that clearer in the PR description.

That being said, in this particular case, the switch from `_image_rodata_start` to `__rom_start_start` does seem to be wrong.
So far, there is no correct replacement ready for the `_image_rodata_X` symbols.
(there is also no real replacement for `_image_rom_start` and `_image_rom_end`, but for `_image_rom_start` we do have a `rom_start` section that are close, but it only matches the start symbol for the rom_start section, not the end symbol).

With regards to bisectability, the plan is that all cleanup needed to prepare for CMake linker feature should comes first, so that they can be taken in independently, as commented here: https://github.com/zephyrproject-rtos/zephyr/pull/36140#issuecomment-858781685 but i'm not there yet.

Finally, thanks for the nice observation and I will surely take a closer look at `_image_rodata_X`.
Also I hope the additional description gives a little better understanding of the direction and progress of the work.

Reply:
That's helpful, thanks. By the way, will this abstraction with scatter and steering files make it possible to resolve this linker issue which shows up with GNU ld on some out of tree architectures?

https://lists.zephyrproject.org/g/devel/message/7765

Reply:
That's helpful, thanks. By the way, will this abstraction with scatter and steering files make it possible to resolve this linker issue which shows up with GNU ld on some out of tree architectures?

https://lists.zephyrproject.org/g/devel/message/7765

Reply:
Just to be clear, I know that GNU ld and armlink are not the same program. I'm asking if the abstraction you're adding within the build system that lives on top of these two linkers could be adapted to solve the problem of GNU ld also changing the names of some symbols in the case of this Renesas system and apparently some others.

Reply:
@mbolivar-nordic I do think the linker script generator will offer some possibilities to address the issue regarding symbols and the Renesas system.

The exact way of doing so will of course require some exploration work, but from the description in the link I think the linker abstraction has a high probability of being able to solve such an issue.